src/code_8069E0C.c

```diff
@@ -0,0 +1,293 @@
+#include "global.h"
+#include "constants/ability.h"
+#include "constants/move_id.h"
+#include "constants/type.h"
+#include "dungeon_entity.h"
+#include "dungeon_global_data.h"
+#include "dungeon_pokemon_attributes.h"
+#include "dungeon_random.h"
+#include "dungeon_util.h"
+#include "pokemon.h"
+#include "weather.h"
+#include "gUnknown_203B46C.h"
+
+extern struct UnkSaveStruct1 *gUnknown_203B46C;
+
+struct unkStruct_80F520C
+{
+    u8 unk0;
+    u8 unk1;
+    s16 unk2;
+};
+
+extern struct unkStruct_80F520C gUnknown_80F520C[];
+
+extern u8 gAvailablePokemonNames[];
+extern u8 gUnknown_202DE58[];
+extern u8 gUnknown_80F51E4[];
+extern u8 *gUnknown_80FCC7C[];
+extern u8 *gUnknown_80FCCAC[];
+extern u8 *gUnknown_80FDCC8[];
+
+extern u8* GetUnformattedTypeString(s16);
+extern void SetMessageArgument(char[], struct DungeonEntity*, u32);
+void sub_80522F4(struct DungeonEntity *r0, struct DungeonEntity *r1, const char r2[]);
+extern void sub_8042900(struct DungeonEntity *r0);
+extern void sub_8042968(struct DungeonEntity *r0);
+extern void sub_806ABAC(struct DungeonEntity *, struct DungeonEntity *);
+extern void sub_8076090(struct DungeonEntity *, struct DungeonEntity *,u32,u32,u32);
+void sub_8041BBC(struct DungeonEntity *r0);
+extern u8 sub_8045888(struct DungeonEntity *);
+extern void sub_806A2BC(struct DungeonEntity *, u8);
+extern void sub_804178C(u32);
+extern void sub_803F508(struct DungeonEntity *);
+extern void sub_804AC20(struct Position *);
+extern u8 sub_8092364(u8);
+extern void sub_803E46C(u32);
+extern void sub_803E708(u32 r0, u32 r1);
+
+void sub_8069E0C(struct DungeonEntity *param_1)
+{
+  s32 index;
+  struct DungeonEntityData *iVar4;
+
+  iVar4 = param_1->entityData;
+  if (HasAbility(param_1, ABILITY_FORECAST)) {
+    iVar4->types[0] = gUnknown_80F520C[GetWeather(param_1)].unk0;
+    iVar4->types[1] = TYPE_NONE;
+  }
+  else {
+    for(index = 0; index < 2; index++)
+      iVar4->types[index] = GetPokemonType(iVar4->entityID, index);
+  }
+  for(index = 0; index < 2; index++)
+    iVar4->abilities[index] = GetPokemonAbility(iVar4->entityID, index);
+  gDungeonGlobalData->unkC = 1;
+}
+
+void TriggerWeatherAbilities(void)
+{
+  struct DungeonEntity *uVar4;
+  s32 iVar5;
+
+  if (gDungeonGlobalData->unkC != 0) {
+    gDungeonGlobalData->unkC = 0;
+    gDungeonGlobalData->negateWeatherEffects = FALSE;
+
+    for(iVar5 = 0; iVar5 < DUNGEON_MAX_POKEMON; iVar5++)
+    {
+      uVar4 = gDungeonGlobalData->allPokemon[iVar5];
+      if (EntityExists(uVar4)) {
+        if (HasAbility(uVar4, ABILITY_DRIZZLE)) {
+            gDungeonGlobalData->startingWeather[4] = 1;
```

Comment:
These can use the weather macros in `weather.h`.

Reply:
Fixed